src/receive_buffer.zig

```diff
@@ -0,0 +1,371 @@
+        const result = buffer.recieve_size - buffer.consume_size;
+        assert(result <= constants.message_size_max);
+        return result;
+    }
+
+    pub fn invalidate(buffer: *ReceiveBuffer, reason: InvalidReason) void {
+        assert(buffer.invalid == null);
+        buffer.recieve_size = 0;
+        buffer.consume_size = 0;
+        buffer.valid_header = false;
+        buffer.valid_body = false;
+        buffer.invalid = reason;
+    }
+
+    /// Advances the parsing state machine.
+    /// Idempotent, but eagerly called whenever available_slice changes.
+    fn validate(buffer: *ReceiveBuffer) void {
+        if (buffer.invalid == null) buffer.validate_header();
+        if (buffer.invalid == null) buffer.validate_body();
+
+        // Only assertions below this line:
+        if (buffer.invalid != null) {
+            assert(!buffer.valid_header);
+            assert(!buffer.valid_body);
+        }
+        if (buffer.valid_body) assert(buffer.valid_header);
+        assert(buffer.consume_size <= buffer.recieve_size);
+    }
+
+    fn validate_header(buffer: *ReceiveBuffer) void {
+        assert(buffer.invalid == null);
+        if (buffer.valid_header) {
+            assert(buffer.available_size() >= @sizeOf(Header));
+            return;
+        }
+        if (buffer.available_size() < @sizeOf(Header)) return;
+
+        const header_bytes = buffer.available_slice()[0..@sizeOf(Header)];
+
+        var header: Header = undefined;
+        stdx.copy_disjoint(.exact, u8, std.mem.asBytes(&header), header_bytes);
+
+        // NB: Avoid materializing invalid Zig enum value.
+        comptime assert(@sizeOf(vsr.Command) == @sizeOf(u8) and
+            std.meta.FieldType(Header, .command) == vsr.Command);
+        const command_raw: u8 = header_bytes[@offsetOf(Header, "command")];
+        _ = std.meta.intToEnum(vsr.Command, command_raw) catch {
+            vsr.fatal(
+                .unknown_command,
+                "unknown command, crashing for safety " ++
+                    "(peer={} command={d} protocol={d} replica={d} release={})",
+                .{
+                    buffer.peer,
+                    command_raw,
+                    header.protocol,
+                    header.replica,
+                    header.release,
+                },
+            );
+        };
+
+        if (header.size < @sizeOf(Header) or header.size > constants.message_size_max) {
+            buffer.invalidate(.header_size);
+            return;
+        }
+        assert(@sizeOf(Header) <= header.size and header.size <= constants.message_size_max);
+
+        if (!header.valid_checksum()) {
+            buffer.invalidate(.header_checksum);
+            return;
+        }
+
+        buffer.valid_header = true;
+
+        // To avoid dropping outgoing messages, set the peer as soon as we can,
+        // and not when we receive a full message.
+        const message_peer = header.peer_type();
```

Comment:
Ideally, I'd love to do this inside Replica.on_messages, so that ReceiveBuffer doesn't have to know about the content of the messages at all, but the problem with that is that, by the time on_messages returns and the bus gets a chance to setup connection's peer, the replica will have already sent response messages!